recipes-security/tboot/tboot-1.9.9/0008-pcr-calc-Add-0x40f-event-emulation.patch

```diff
@@ -0,0 +1,930 @@
+predictions of these values, introduce the scafolding to support such
+emulation and allow the pcr-calc util to forward seal against a
+different version of TBoot using a possibly different ACM.
+
+Add 2 options to pcr-calc and modify existing `-e`:
+    -e evt_type:<hash_str>|emulate Override PCR events type found in the
+       logs with the provided value or emulate the value entirely
+       (emulation requires TBoot version and ACM to be provided).
+
+    -A acm-path Path to the ACM that will be used to emulate events for
+       PCR calculations (see -e).
+
+    -V tboot-version TBoot version targeted to emulate events for PCR
+       calculations (see -e).
+
+Signed-off-by: Eric Chanudet <chanudete@ainfosec.com>
+---
+ pcr-calc/Makefile   |   2 +-
+ pcr-calc/acm.c      | 216 ++++++++++++++++++++++++++++++++++++++++++++
+ pcr-calc/acm.h      | 134 +++++++++++++++++++++++++++
+ pcr-calc/eventlog.c | 175 ++++++++++++++++++++++++++++++++++-
+ pcr-calc/eventlog.h |  18 ++++
+ pcr-calc/pcr-calc.c |  50 +++++++++-
+ pcr-calc/tpm.c      |  48 +++++++++-
+ pcr-calc/tpm.h      |  11 +++
+ pcr-calc/util.c     |  30 +++++-
+ pcr-calc/util.h     |   5 +-
+ 10 files changed, 678 insertions(+), 11 deletions(-)
+ create mode 100644 pcr-calc/acm.c
+ create mode 100644 pcr-calc/acm.h
+
+diff --git a/pcr-calc/Makefile b/pcr-calc/Makefile
+index e49102d..7b06c5a 100644
+--- a/pcr-calc/Makefile
++++ b/pcr-calc/Makefile
+@@ -60,7 +60,7 @@ distclean : clean
+ BUILD_DEPS := $(ROOTDIR)/Config.mk $(CURDIR)/Makefile
+ 
+ 
+-pcr-calc : eventlog.o pcr-calc.o tpm.o util.o hash.o tb_policy.o
++pcr-calc : acm.o eventlog.o pcr-calc.o tpm.o util.o hash.o tb_policy.o
+ 	$(CC) $(CFLAGS) $(LDFLAGS) $^ $(LIBS) -o $@
+ 
+ module_hash : module_hash.o hash.o
+diff --git a/pcr-calc/acm.c b/pcr-calc/acm.c
+new file mode 100644
+index 0000000..56e76b2
+--- /dev/null
++++ b/pcr-calc/acm.c
+@@ -0,0 +1,216 @@
++#include <stdio.h>
++#include <stdlib.h>
++#include <errno.h>
++#include <unistd.h>
++#include <fcntl.h>
++#include <string.h>
++#include <inttypes.h>
++#include <sys/mman.h>
++#include <sys/stat.h>
++
++#include "uuid.h"
++#include "acm.h"
++
++#define DEBUG 0
++#define printd(fmt, ...)					\
++	if (DEBUG) {						\
++		fprintf(stdout, fmt "\n", ##__VA_ARGS__);	\
++	}
++
++static acm_hdr_t *get_acm_header(void *p)
++{
++	uint32_t ovf;
++	acm_hdr_t *hdr = p;
++
++	if (__builtin_umul_overflow(hdr->size, 4, &ovf))
++		return NULL;
++	if (hdr->module_type != ACM_TYPE_CHIPSET)
++		return NULL;
++	if (hdr->module_vendor != ACM_VENDOR_INTEL)
++		return NULL;
++
++	return hdr;
++}
++
++static acm_info_table_t *get_acm_info_table(const acm_hdr_t *hdr)
++{
++	uint32_t ovf;
++	uint32_t offset;
++	acm_info_table_t *infotable;
++
++	/* Check infotable is in ACM. */
++	if (__builtin_uadd_overflow(hdr->header_len, hdr->scratch_size, &ovf))
++		return NULL;
++	if (__builtin_umul_overflow(hdr->header_len + hdr->scratch_size, 4, &ovf))
```

Comment:
__builtin_uadd_overflow stores the result into ovf, so you can do `__builtin_umul_overflow(ovf, 4, &ovf)` and then use the new ovf again `ovf > (hdr->size * 4)`

Reply:
Addressed in v3.